Apply Redux

kuhiepmr opened the pull request:
- Add Reducer
- Add Action
- Update components following redux concept

**GIF**
![codenote-#2](https://user-images.githubusercontent.com/61727097/75942396-db0ec280-5ec4-11ea-8ddf-c91a1aa7cc43.gif)

kuhiepmr committed:
Apply Redux

- Add Reducer
- Add Action
- Update components following redux concept

huynvk commented on src/components/AuthenticatedRoute/index.tsx:
```diff
@@ -1,14 +1,21 @@
+import { Redirect, Route } from "react-router-dom";
 
-export default ({ component: C, props: cProps, ...rest }) =>
+const Authenticated = ({ component: C, isAuthenticated, ...rest }) =>
```
`Authenticated` is an adjective, not a good name for react component

kuhiepmr replied:
Sorry, I quick named it because I let it export default with the file name
Fixed

huynvk commented on src/components/Navigation/index.tsx:
```diff
@@ -40,4 +42,12 @@ class Navigation extends React.Component {
 
-export default Navigation;
+const mapStateToProps = (state) => ({
+  isAuthenticated: state.authenticate.isAuthenticated,
```
I see this prop `isAuthenticated` is being used at many places with duplicated redux code. You can consider adding a HOC for this instead (e.g. `withAuthenticationStatus`)

kuhiepmr replied:
Updated

huynvk commented on src/components/Navigation/index.tsx:
```diff
@@ -40,4 +42,12 @@ class Navigation extends React.Component {
+});
+
+const mapDispatchToProps = dispatch => ({
+  userHasAuthenticated: (bool) => dispatch(authenticate(bool)),
```
Same here, you can add HOC `withAuthenticate` to remove duplications.

kuhiepmr replied:
Updated

huynvk commented on src/components/UnauthenticatedRoute/index.tsx:
```diff
@@ -17,19 +18,23 @@ const querystring = (name, url = window.location.href) => {
 
-export default UnauthenticatedRoute;
+const mapStateToProps = (state) => ({
+  isAuthenticated: state.authenticate.isAuthenticated,
```
Use HOC as suggested above

kuhiepmr replied:
Updated

huynvk commented on src/components/Navigation/index.tsx:
```diff
@@ -40,4 +42,12 @@ class Navigation extends React.Component {
+  userHasAuthenticated: (bool) => dispatch(authenticate(bool)),
+})
+
+export default connect(mapStateToProps, mapDispatchToProps)(Navigation);
```
In case you use many HOC, you can apply `compose` for more readable.

https://reactjs.org/docs/higher-order-components.html#convention-maximizing-composability

kuhiepmr replied:
Updated

huynvk commented on .vscode/settings.json:
```diff
@@ -0,0 +1 @@
+{}
```
I thought this file is ignored?

kuhiepmr replied:
I missed it on this branch 🤣

huynvk commented on src/actions/authenticate.tsx:
```diff
@@ -0,0 +1,8 @@
+
+export const authenticate = (payload) => ({
+  type: types.AUTHENTICATE,
+  payload
```
Use the payload blindly as provided from outside introduce risks and hard to use later on. After one month, you won't remember what payload should look like. You can rewrite it a bit:

```js
export const authenticate = (isAuthenticate) => ({
  type: types.AUTHENTICATE,
  payload: {
    isAuthenticate
  }
})
```

huynvk replied:
Actually, it's better if you have 2 action for login and logout without any parameter here, the producer should handle it correctly base on action type
```js
export const login = () => ({ type: types.LOGIN })
export const logout = () => ({ type: types.LOGOUT })
```

huynvk replied:
You can also use boilerplate for them as well. E.g. https://github.com/redux-utilities/redux-actions

huynvk commented on src/components/AuthenticatedRoute/index.tsx:
```diff
@@ -1,14 +1,21 @@
-              .search}`}
-          />}
+          to={`/login?redirect=${props.location.pathname}${props.location
+            .search}`}
```
Line wrap here seems strange. Or you can:
```js
const { pathname, search } = props.location

...

to={`/login?redirect=${pathname}${search}`}
```

huynvk commented on src/components/Navigation/index.tsx:
```diff
@@ -25,13 +27,13 @@ class Navigation extends React.Component {
+                <LinkContainer to="/login">
+                  <NavItem>Login</NavItem>
+                </LinkContainer>
+              </Fragment>
```
It seems the previous indent more correct, doesn't it?

kuhiepmr replied:
Because of the IDE 😄

huynvk commented on src/components/UnauthenticatedRoute/index.tsx:
```diff
@@ -17,19 +18,23 @@ const querystring = (name, url = window.location.href) => {
-              to={redirect === "" || redirect === null ? "/" : redirect}
-            />}
+            to={redirect === "" || redirect === null ? "/" : redirect}
+          />}
```
The indent seems wrong. Is it your IDE or you reformat it yourself?

kuhiepmr replied:
Because of the IDE 😄

huynvk commented on src/index.tsx:
```diff
@@ -1,11 +1,13 @@
+import config from './config';
+import './index.css';
 import * as serviceWorker from './serviceWorker';
+import store from './store';
```
We can follow import order convention here:
- External library (react, react-router-dom, etc.)
- Component from other package inside project
- Component within the folder
- Css

kuhiepmr replied:
Fixed

huynvk commented on src/reducers/authenticate.ts:
```diff
@@ -1,6 +1,17 @@
+      return {
+        ...state,
+        isAuthenticated: action.payload
+      };
```
As I suggested above, we can divide it to 2 actions LOGIN and LOGOUT:

```js
switch (action.type) {
    case types.LOGIN:
      return { ...state, isAuthenticated: true };
    case types.LOGOUT:
      return { ...state, isAuthenticated: false };
    ...
}
```

huynvk replied:
You can apply boilerplate as well

kuhiepmr replied:
Updated

huynvk commented on src/screens/Home/index.tsx:
```diff
@@ -38,25 +40,25 @@ class Home extends Component {
+            <ListGroupItem header={note.content.trim().split("\n")[0]}>
+              {"Created: " + new Date(note.createdAt).toLocaleString()}
+            </ListGroupItem>
+          </LinkContainer>
```
The indent looks wrong

kuhiepmr replied:
Because of the IDE  😄

huynvk replied:
Can you update your IDE and correct the file?

huynvk commented on src/screens/Home/index.tsx:
```diff
@@ -87,4 +89,12 @@ class Home extends Component {
 
-export default Home;
+const mapStateToProps = (state) => ({
+  isAuthenticated: state.authenticate.isAuthenticated,
```
Can use HOC instead

kuhiepmr replied:
Updated

huynvk commented on src/screens/Home/index.tsx:
```diff
@@ -87,4 +89,12 @@ class Home extends Component {
+});
+
+const mapDispatchToProps = dispatch => ({
+  userHasAuthenticated: (bool) => dispatch(authenticate(bool)),
```
This one can be replaced by a HOC as well

kuhiepmr replied:
Updated

huynvk commented on src/store.tsx:
```diff
@@ -0,0 +1,7 @@
+
+export default createStore(
+  rootReducer,
+  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
```
This one is for debug only. You can either remove it or put a comment here (e.g. // TODO: remove this line before deployment)

kuhiepmr replied:
Sorry, I forgot to remove it
Fixed

kuhiepmr committed:
Update redux

- Separate AUTHENTICATE action types -> LOGIN/LOGOUT
- Use HOC instead of connect() directly
- Remove DEV mode in store

huynvk commented on src/actions/authenticate.tsx:
```diff
@@ -1,8 +1,11 @@
+ */
+export const authenticate = (isLogin) => ({
+  type: isLogin ? types.LOGIN : types.LOGOUT,
+})
```
I would recommend to write separate action creators for login and logout
